File: src/lib/api/image_tag.ts
import { invoke } from '@tauri-apps/api';

export interface ImageTag {
	relative_path: string;
	source_id: string;
	tag_id: string;
}

export interface ImageTags {
	image_tags: ImageTag[];
}

export interface TagAssignments {
	add: string[];
	remove: string[];
}

export const create = async (name: string, parentId: string): Promise<ImageTag> => {
	const imageTag = (await invoke('create_image_tag', { name, parentId })) as ImageTag;

	return imageTag;
};

export const list = async (): Promise<ImageTags> => {
	const imageTags = (await invoke('list_image_tags', {})) as ImageTags;

	return imageTags;
};

export const listByRelativePath = async (
	relativePath: string,
	sourceId: string
): Promise<ImageTags> => {
	const imageTags = (await invoke('list_image_tags_by_relative_path', {
		relativePath,
		sourceId
	})) as ImageTags;

	return imageTags;
};

export const get = async (id: string): Promise<ImageTag> => {
	const imageTag = (await invoke('get_image_tag', { id })) as ImageTag;

	return imageTag;
};

export const remove = async (
	tagId: string,
	sourceId: string,
	relativePath: string
): Promise<ImageTag> => {
	const imageTag = (await invoke('delete_image_tag', {
		tagId,
		sourceId,
		relativePath
	})) as ImageTag;

	return imageTag;
};

export const assign = async (
	relativePaths: string[],
	sourceId: string,
	assignments: TagAssignments
) => {
	await invoke('assign_image_tags', {
		relativePaths,
		sourceId,
		assignments
	});
};

File: src-tauri/src/cmd/image_tag.rs
use crate::{
    api::image_tag::{ImageTags, TagAssignments},
    data::image_tag::entity,
    errors::AppError,
    state::ServiceAccess,
};

#[tauri::command]
pub fn create_image_tag(
    tag_id: &str,
    relative_path: &str,
    source_id: &str,
    handle: tauri::AppHandle,
) -> Result<entity::ImageTag, AppError> {
    handle.image_tag_ctrl(|ctrl| ctrl.create(tag_id, relative_path, source_id))
}

#[tauri::command]
pub fn get_image_tag(
    tag_id: &str,
    relative_path: &str,
    source_id: &str,
    handle: tauri::AppHandle,
) -> Result<entity::ImageTag, AppError> {
    handle.image_tag_ctrl(|ctrl| ctrl.get_by_ids(tag_id, relative_path, source_id))
}

#[tauri::command(async)]
pub fn list_image_tags(handle: tauri::AppHandle) -> Result<ImageTags, AppError> {
    handle
        .image_tag_ctrl(|ctrl| ctrl.list())
        .map(ImageTags::from)
}

#[tauri::command(async)]
pub fn list_image_tags_by_relative_path(
    relative_path: &str,
    source_id: &str,
    handle: tauri::AppHandle,
) -> Result<ImageTags, AppError> {
    handle
        .image_tag_ctrl(|ctrl| ctrl.list_by_relative_path(relative_path, source_id))
        .map(ImageTags::from)
}

#[tauri::command]
pub fn delete_image_tag(
    tag_id: &str,
    relative_path: &str,
    source_id: &str,
    handle: tauri::AppHandle,
) -> Result<(), AppError> {
    handle.image_tag_ctrl(|ctrl| ctrl.delete(tag_id, relative_path, source_id))
}

#[tauri::command]
pub fn assign_image_tags(
    relative_paths: Vec<String>,
    source_id: &str,
    assignments: TagAssignments,
    handle: tauri::AppHandle,
) -> Result<(), AppError> {
    handle.image_tag_ctrl(|ctrl| ctrl.assign(relative_paths, source_id, assignments))
}
